src/cls/queue/cls_queue_src.cc

```diff
@@ -0,0 +1,901 @@
+
+  bl.claim_append(bl_head);
+
+  CLS_LOG(0, "INFO: cls_create_queue writing head of size %u", bl.length());
+  ret = cls_cxx_write2(hctx, 0, bl.length(), &bl, CEPH_OSD_OP_FLAG_FADVISE_WILLNEED);
+  if (ret < 0) {
+    CLS_LOG(0, "ERROR: %s(): cls_cxx_write returned %d", __func__, ret);
+    return ret;
+  }
+  return 0;
+}
+
+int cls_init_queue(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  //get head and its size
+  uint64_t head_size = 0;
+  cls_queue_head head;
+  int ret = get_queue_head_and_size(hctx, head, head_size);
+
+  //head is already initialized
+  if (ret == 0) {
+    CLS_LOG(1, "INFO: cls_init_queue_op(): head is already initialized\n");
+    return 0;
+  }
+
+  if (ret < 0 && ret != -EINVAL) {
+    return ret;
+  }
+
+  auto in_iter = in->cbegin();
+  cls_init_queue_op op;
+  try {
+    decode(op, in_iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_init_queue_op(): failed to decode entry\n");
+    return -EINVAL;
+  }
+
+  CLS_LOG(10, "INFO: cls_init_queue_op init queue of head size %lu", op.head_size);
+  CLS_LOG(10, "INFO: cls_init_queue_op init queue of size %lu", op.head.queue_size);
+
+  head_size = QUEUE_HEAD_SIZE_1K;
+
+  if (op.head.has_urgent_data) {
+    if (op.head_size == 0) {
+      head_size = QUEUE_HEAD_SIZE_4K;
+      op.head.tail = op.head.front = QUEUE_START_OFFSET_4K;
+      op.head.last_entry_offset = QUEUE_START_OFFSET_4K;
+    } else {
+      head_size = op.head_size;
+      op.head.tail = op.head.front = head_size;
+      op.head.last_entry_offset = head_size;
+    }
+  } else {
+    head_size = QUEUE_HEAD_SIZE_1K;
+    op.head.tail = op.head.front = QUEUE_START_OFFSET_1K;
+    op.head.last_entry_offset = QUEUE_START_OFFSET_1K;
+  }
+  op.head.queue_size += head_size;
+
+  CLS_LOG(10, "INFO: cls_init_queue_op queue actual size %lu", op.head.queue_size);
+  CLS_LOG(10, "INFO: cls_init_queue_op head size %lu", head_size);
+  CLS_LOG(10, "INFO: cls_init_queue_op queue front offset %lu", op.head.front);
+
+  //encode head size
+  bufferlist bl;
+  encode(head_size, bl);
+  CLS_LOG(0, "INFO: cls_init_queue_op head size %u", bl.length());
+
+  //encode head
+  bufferlist bl_head;
+  encode(op.head, bl_head);
+
+  bl.claim_append(bl_head);
+
+  CLS_LOG(0, "INFO: cls_init_queue_op writing head of size %u", bl.length());
+  ret = cls_cxx_write2(hctx, 0, bl.length(), &bl, CEPH_OSD_OP_FLAG_FADVISE_WILLNEED);
+  if (ret < 0) {
+    CLS_LOG(0, "ERROR: %s(): cls_init_queue_op returned %d", __func__, ret);
+    return ret;
+  }
+  return 0;
+}
+
+int cls_get_queue_size(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  //get head and its size
+  uint64_t head_size = 0;
+  cls_queue_head head;
+  int ret = get_queue_head_and_size(hctx, head, head_size);
+  if (ret < 0) {
+    return ret;
+  }
+
+  head.queue_size -= head_size;
+
+  CLS_LOG(10, "INFO: cls_get_queue_size: size of queue is %lu\n", head.queue_size);
+
+  encode(head.queue_size, *out);
```

Comment:
as trivial as this op is, i still think it deserves a versioned `struct cls_queue_get_size_ret` so we have the option to change/extend it in the future

Reply:
Ok